Content.Server/Doors/Systems/DoorSystem.cs

```diff
@@ -1,69 +1,293 @@
+    ///     is holding the correct tool.
+    /// </summary>
+    private async void TryWeldDoor(EntityUid target, EntityUid used, EntityUid user, DoorComponent door)
+    {
+        if (!door.Weldable || door.BeingWelded || door.CurrentlyCrushing.Count > 0)
+            return;
+
+        // is the door in a weld-able state?
+        if (door.State != DoorState.Closed && door.State != DoorState.Welded)
+            return;
+
+        // perform a do-after delay
+        door.BeingWelded = true;
+        _toolSystem.UseTool(used, user, target, 3f, 3f, door.WeldingQuality,
+            new WeldFinishedEvent(), new WeldCancelledEvent(), target);
+    }
+
+    /// <summary>
+    ///     Pry open a door. This does not check if the user is holding the required tool.
+    /// </summary>
+    private async void TryPryDoor(EntityUid target, EntityUid tool, EntityUid user, DoorComponent door)
+    {
+        if (door.State == DoorState.Welded)
+            return;
+
+        var canEv = new BeforeDoorPryEvent(user);
+        RaiseLocalEvent(target, canEv, false);
+
+        if (canEv.Cancelled)
+            return;
+
+        var modEv = new DoorGetPryTimeModifierEvent();
+        RaiseLocalEvent(target, modEv, false);
+
+        _toolSystem.UseTool(tool, user, target, 0f, modEv.PryTimeModifier * door.PryTime, door.PryingQuality,
+                new PryFinishedEvent(), new PryCancelledEvent(), target);
+    }
+
+    private void OnWeldCancelled(EntityUid uid, DoorComponent door, WeldCancelledEvent args)
+    {
+        door.BeingWelded = false;
+    }
+
+    private void OnWeldFinished(EntityUid uid, DoorComponent door, WeldFinishedEvent args)
+    {
+        door.BeingWelded = false;
+
+        if (!door.Weldable)
+            return;
+
+        if (door.State == DoorState.Closed)
+            SetState(uid, DoorState.Welded, door);
+        else if (door.State == DoorState.Welded)
+            SetState(uid, DoorState.Closed, door);
+    }
+
+    private void OnPryCancelled(EntityUid uid, DoorComponent door, PryCancelledEvent args)
+    {
+        door.BeingPried = false;
+    }
+
+    private void OnPryFinished(EntityUid uid, DoorComponent door, PryFinishedEvent args)
+    {
+        door.BeingPried = false;
+
+        if (door.State == DoorState.Closed)
+            StartOpening(uid, door);
+        else if (door.State == DoorState.Open)
+            StartClosing(uid, door);
+    }
+    #endregion
+
+    /// <summary>
+    ///     Does the user have the permissions required to open this door?
+    /// </summary>
+    public override bool HasAccess(EntityUid uid, EntityUid? user = null, AccessReaderComponent? access = null)
+    {
+        // TODO network AccessComponent for predicting doors
+
+        // if there is no "user" we skip the access checks. Access is also ignored in some game-modes.
+        if (user == null || AccessType == AccessTypes.AllowAll)
+            return true;
+
+        if (!Resolve(uid, ref access))
+            return true;
```

Comment:
I get resolve exceptions trying to open doors without accessreader so this should probably be false?

Reply:
yeah it should